src/KubernetesClient/Controllers/ResourceEventDeltaBlock.cs

```diff
@@ -0,0 +1,304 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reactive.Disposables;
+using System.Threading;
+using System.Threading.Tasks;
+using System.Threading.Tasks.Dataflow;
+using k8s.Informers.Notifications;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Abstractions;
+
+namespace k8s.Controllers
+{
+    /// <summary>
+    ///     A TPL Dataflow block exposes queuing semantics that attaches to infromer observable streams.
+    ///     It groups delta changes per key, sending each group to downstream blocks as individual messages
+    ///     This allows "Per resource ID" processing semantics
+    ///     Any sync events are only propagated if no other changes are queued up
+    /// </summary>
+    /// <typeparam name="TKey">The type of key used to identity resource</typeparam>
+    /// <typeparam name="TResource">The resource type</typeparam>
+    public class ResourceEventDeltaBlock<TKey, TResource> : ISourceBlock<List<ResourceEvent<TResource>>>, ITargetBlock<ResourceEvent<TResource>>
```

Comment:
Do we need a dependency on dataflow? This library is extremely allocation inefficient.

Reply:
I chose TPL dataflow because when you deal with controllers, it exposes useful queue like semantics which are intuitive. I could model this with RX, but to gain same functionality we're gonna need two observables one that is supplying data, and another to mark boundaries of when worker is ready for more. This will end up very similar in syntax to RX Window function, which to most dev not familiar with RX goes right over their head.

I brought this up a number of times, but I actually want to get blessing to modularize the SDK, and then TPL specific stuff around workflow scheduling for controllers would be moved there. I see a need for at least 3 extra packages:
KubernetesClient.Informers - all the RX stuff
KubernetesClient.Controllers - all the TPL stuff, a new pieces of code I plan to add to create reusable reconciliation plans
KubernetesClient.DependencyInjection - stuff to add all the extension methods into like the one currently in sample project

Reply:
I originally had it coded with System.Threading.Channels, but switched to TPL Dataflow as it gives a fine-grained control over "number of concurrent workers" semantics.

Reply:
I think you should switch it back.

Reply:
Removed this code for now, will reimplement on top of RX without TPL Dataflow once more design patterns for .NET operators are flushed out.